addons/mass_mailing_sms/models/mailing_list.py

```diff
@@ -63,3 +63,106 @@ def _mailing_get_opt_out_list_sms(self, mailing):
+        """
+        # Explation of the SQL query with an example. There are the following lists
+        # A (id=4): yti@odoo.com; yti@example.com
+        # B (id=5): yti@odoo.com; yti@openerp.com
+        # To merge the mailing lists A and B, we build the view st that looks
+        # like this with our example:
+        #
+        #  contact_id |           email           | row_number |  list_id |
+        # ------------+---------------------------+------------------------
+        #           4 | yti@odoo.com              |          1 |        4 |
+        #           6 | yti@odoo.com              |          2 |        5 |
+        #           5 | yti@example.com           |          1 |        4 |
+        #           7 | yti@openerp.com           |          1 |        5 |
+        #
+        # The row_column is kind of an occurence counter for the email address.
+        # Then we create the Many2many relation between the destination list and the contacts
+        # while avoiding to insert an existing email address (if the destination is in the source
+        # for example)
+        self.ensure_one()
+        # Put destination is sources lists if not already the case
+        src_lists |= self
+        self.env['mailing.contact'].flush(['email', 'email_normalized'])
+        self.env['mailing.contact.subscription'].flush(['contact_id', 'opt_out', 'list_id'])
+        self.env.cr.execute("""
+            INSERT INTO mailing_contact_list_rel (contact_id, list_id)
+            SELECT st.contact_id AS contact_id, %s AS list_id
+            FROM
+                (
+                SELECT
+                    contact.id AS contact_id,
+                    contact.email AS email,
+                    list.id AS list_id,
+                    row_number() OVER (PARTITION BY email ORDER BY email) AS rn
+                FROM
+                    mailing_contact contact,
+                    mailing_contact_list_rel contact_list_rel,
+                    mailing_list list
+                WHERE contact.id=contact_list_rel.contact_id
+                AND COALESCE(contact_list_rel.opt_out,FALSE) = FALSE
+                AND (contact.email_normalized NOT IN (select email from mail_blacklist where active = TRUE)
+                OR contact.mobile IS NOT NULL)
+                AND list.id=contact_list_rel.list_id
+                AND list.id IN %s
+                AND NOT EXISTS
+                    (
+                    SELECT 1
+                    FROM
+                        mailing_contact contact2,
+                        mailing_contact_list_rel contact_list_rel2
+                    WHERE contact2.email = contact.email
+                    AND contact_list_rel2.contact_id = contact2.id
+                    AND contact_list_rel2.list_id = %s
+                    )
+                ) st
+            WHERE st.rn = 1;""", (self.id, tuple(src_lists.ids), self.id))
+        self.env.cr.execute("""
+            UPDATE mailing_contact_list_rel
```

Comment:
Unsure about this one ... can't we simply do it in a single query by correctly propagating opt-out in the first query ? Is there any issue preventing to do it ?

I suppose merging opt-out coming from several lists is done like "at least one opt-out means opt-out" ? in which case maybe a coalesce could be used.

Reply:
Hi,

As trying to make it in a single queries took too much time and didn't seem possible to me (either I had a problem with duplicates or opt_out), I've decided to go back to 2 queries to finish this task.

Reply:
@tde-banana-odoo I've been thinking about this, I don't think there can be a much more efficient solution because we choose to keep the rows of the biggest relation. 

We could create a view for the join, add a constraint on the view, and use ON CONFLICT but the cost of the view seems prohibitive. MERGE also allows conditional INSERT/UPDATE but it doesn't really work for this usecase.

I don't think this can be improved in this way, I'll try improving the current approach if I can though.